---
title: 'Data Preperation for Bird Species in Comparative Analysis'
author: "Justin G. Cally"
subtitle: Supplementary Material
output:
  html_document:
    toc: true # table of content true
    toc_float: false # make 
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: yeti # lovely fonts and colours
    code_folding: hide # awesome buttons to show/hide the code
  pdf_document: default
---

# Bird Environmental Data

## Setup  
Load in packages used in the analysis
```{r, warning=FALSE, message=FALSE}
library(raster)
library(sp)
#devtools::install_github("r-spatial/sf") #install gdal via brew
library(sf)
library(rgdal)
library(maptools)
library(pander)
library(plyr)
library(dplyr)
library(tidyverse)
library(knitr)
library(lsr)
library(data.table)
library(readxl)
library(fuzzyjoin)
library(parallel)
# devtools::install_github("jimhester/archive") # Need to use: brew install libarchive in bash
# devtools::install_github("mirzacengic/climatedata")
library(climatedata)
options(stringsAsFactors = FALSE)
```

## Read in all species

Because the HBW taxonomy changes year to year we will have to use a field known as the SpcRecID and the SISRecID. SpcRecID was used pre-2014 and thus is used for the birdtree.org phylogeny. SISRecID is used post 2014 and is thus used for this edition of the spatial dataset. 

```{r data_read}
if(!file.exists("data/Intermediate_data/BOTW.rds")){
all.data <- sf::read_sf("data/BOTW/BOTW.gdb") #read_sf is supposed to be faster than readOGR
saveRDS(all.data, "data/Intermediate_data/BOTW.rds")
} else {
  all.data <- readRDS("data/Intermediate_data/BOTW.rds")
  }
nest.data <- read_csv("data/Combined_allspp.csv", ) %>% mutate(SCINAME = str_replace_all(species, "_", " "))

# HBW checklists from 2010 (when Jetz made phylogeny) and most recent for shapefiles
HBW.checklist.2010 <- read_excel("taxa_details/BirdLife_Checklist_Version_3.xls") %>% select(`Scientific name`, contains("Synonyms"), contains("SpcRecID")) %>% `colnames<-`(c("SCINAME", "Synonyms", "SpcRecID"))
HBW.checklist.2018 <- read_excel("taxa_details/HBW-BirdLife_Checklist_Version_3.xlsx", skip = 1) %>% select(`Scientific name`, contains("SpcRecID"), contains("SISRecID")) %>% `colnames<-`(c("SCINAME", "SpcRecID", "SISRecID"))

# Read in the metadata for the Jetz phylogeny. We should use these names in the phylogenetic analysis
Jetz.Species <- read.csv("taxa_details/Jetz_Species.csv") %>% select(WJSpecID, Scientific)

# Read in the data that had non-matching taxanomic names/missing species ID
missing.species.id <- read.csv("taxa_details/no_SpcRecID_IM.csv") %>% 
  mutate(SCINAME = str_replace_all(species, "_", " ")) %>% 
  select(SpcRecID.new = SpcRecID, SCINAME) %>% na.omit()
jetz.taxonomy.missing <- read.csv("taxa_details/Jetz_taxonomy_missing.csv")
```

## Deal with missing data

Using some fuzzy string matching format the spatial data in a more complete fashion and using the SISID. Range size is also calculated here.

```{r missingdata}
#### Data match to phylogeny ####
# Group SpecRecID as a vector column as some species have multiple IDs. 
HBW.checklist.2010.collapsed <- HBW.checklist.2010 %>%
  filter(!is.na(SCINAME))
  # group_by(SCINAME) %>%
  # summarise(SpcRecID = I(list(SpcRecID)))
# sum(is.null(HBW.checklist.2010.collapsed$SpcRecID))
# # Using the nest data add the Species IDs (matching by SCINAME and also synonyms)
nest.data.withID <- nest.data %>% left_join(HBW.checklist.2010.collapsed)
# length(which(sapply(nest.data.withID$SpcRecID, is.null)))
# # 56
# # Match by synonyms
nest.data.withID2 <- HBW.checklist.2010 %>% select(-SCINAME, Syn.SpcRecID = SpcRecID) %>%
  fuzzy_right_join(nest.data.withID, by = c("Synonyms" = "SCINAME"), match_fun = str_detect) %>% 
  mutate(SpcRecID = coalesce(SpcRecID, Syn.SpcRecID))
  # mutate(SpcRecID = case_when(!is.na(Syn.SpcRecID) ~ pmap(list(SpcRecID, Syn.SpcRecID), c),
  #                             is.na(Syn.SpcRecID) ~ pmap(list(SpcRecID), c)))
# length(which(sapply(nest.data.withID2$SpcRecID, is.null)))
# # 44

# # Add in the missing species ~ 40 species 
nest.data.withID3 <- nest.data.withID2 %>% left_join(missing.species.id) %>%
  mutate(SpcRecID = coalesce(SpcRecID.new, as.integer(SpcRecID)))
#   mutate(SpcRecID = case_when(!is.na(SpcRecID.new) ~ pmap(list(SpcRecID, SpcRecID.new), c),
#                               is.na(SpcRecID.new) ~ pmap(list(SpcRecID), c)))
# length(which(sapply(nest.data.withID3$SpcRecID, is.null)))
# # 12 (removed)

nest.data.jetz <- nest.data.withID3 %>% 
  left_join(jetz.taxonomy.missing %>% select(SCINAME, sp_tax_jetz)) %>%
  mutate(SCINAME = coalesce(sp_tax_jetz, SCINAME)) %>%
  left_join(Jetz.Species, by = c("SCINAME" = "Scientific")) %>%
  filter(!is.na(WJSpecID)) # REMOVE 3 SPECIES NOT INCLUDED IN JETZ

#### Data match to shapefiles ####
species.with.shapes <- all.data %>% as.data.frame() %>% select(SISID) %>% unique()

#### Species to extract data for ####
nest.2018.ID.data <- nest.data.jetz %>% 
  inner_join(HBW.checklist.2018 %>% rename(SCINAME2018 = SCINAME), by = "SpcRecID") %>%
  filter(!is.na(SpcRecID)) %>%
  group_by(SISRecID) %>%
  summarise(SpcRecID = list(unique(SpcRecID)),
            SCINAME2018 = list(unique(SCINAME2018)),
            Jetz.name = list(unique(SCINAME))) %>%
  ungroup()

if(!file.exists("data/Intermediate_data/bird.ranges.breeding.Rds")){
bird.ranges <- all.data %>% 
  filter(SISID %in% nest.2018.ID.data$SISRecID & SEASONAL %in% c(1,2)) %>% #Resident or breeding ranges
  st_cast("MULTIPOLYGON") %>% #Remove MULTISURFACE shapes that ruin everything
  group_by(SISID) %>%
  summarise(SCINAME = glue::glue_collapse(SCINAME, sep = ", "), 
            SEASONAL = glue::glue_collapse(SEASONAL, sep = ", "),
            Shape = st_union(Shape), #combine the multiple polygons for each species
            Range.Size = st_area(Shape))
saveRDS(bird.ranges, file = "data/Intermediate_data/bird.ranges.breeding.Rds")
} else {
  bird.ranges <- readRDS("data/Intermediate_data/bird.ranges.breeding.Rds")
}

```

## Download Worldclim data

Using the climatedata package extract worldclim data: we use grids of 2.5 min.
```{r worldclim_data}
### Get Data Chelsea for the current period
chelsa_bioclim_current <- get_chelsa(type = "bioclim",
                                     output_dir = "data/climate",
                                     period = "current") #Downloads first time but then reads in tifs

# It's now saved in the data folder

# FUTURE CLIMATES #
### Get Future climate predictions: Run this later
# Don't think this is required unless you want to predict future

# From these models we want to narrow it down to a subset that has low interdependency; this means we get good estimates of uncertainty without having to extract data from all the models. This ranking is taken from: Sanderson, B.M., Knutti, R. & Caldwell, P. (2015) A Representative Democracy to Reduce Interdependency in a Multimodel Ensemble. Journal of Climate, 28, 5171â€“5194)
# top5.low.interdependent.models <- c("CESM-1-BCG",
#                                   "MPI-ESM-MR",
#                                   "MIROC5",
#                                   "CMCC-CM",
#                                   "CESM1-CAM5")


# Get models with all 4 RCP scenarios
# models_all_rcp <- check_models() %>% 
# group_by(model) %>%
#   mutate(n = n()) %>%
#   ungroup() %>%
#   filter(n == 4) %>%
#   distinct(model) %>%
#   pull()
# if(!dir.exists("data/climate/current")){
# chelsa_bioclim_future <- get_chelsa(output_dir = my_output_directory, 
#                                       period = "future", 
#                                       future_years = "2041-2060", 
#                                       scenario_string = "rcp85", #Get multiple and interpolate
#                                       model_string = models_all_rcp)
# }
```

# Sample ranges

For each range, sample it 1000 times at random. 
```{r sample_ranges, warning=FALSE, message=FALSE, results='hide'}
if(!file.exists("data/intermediate_data/bird.points.rds")){
#Increse the iterations (defaukt is 4) so we can obtain complete samples of each range
bird.range.list <- setNames(split(bird.ranges, 
                                  seq(nrow(bird.ranges))), 
                            bird.ranges$SISID)
bird.points <- list()
bird.points <- mclapply(bird.range.list, function(x){
  sp::spsample(sf::as_Spatial(x), n = 1000, type = "random", iter = 30, quiet = T)
}, mc.cores = 4)
saveRDS(bird.points, "data/intermediate_data/bird.points.rds")
} else {
bird.points <- readRDS("data/intermediate_data/bird.points.rds")
}
```

# Extract cllimate data from ranges
Now we extract the bioclim data from each of the points
```{r extract}
if(!file.exists("data/intermediate_data/bird.values.rds")){
bird.values <- mclapply(bird.points, function(x) {
  raster::extract(chelsa_bioclim_current, x)
  }, mc.cores = 4) #raster cant deal with sf 

saveRDS(bird.values, "data/intermediate_data/bird.values.rds")
} else {
bird.values <- readRDS("data/intermediate_data/bird.values.rds")
}
```

# Get summary output and format with the original data

```{r format_output}
if(!file.exists("data/intermediate_data/bird.data.recombined.rds")){
bird.frame <- mclapply(bird.values, function(x) {as.data.frame(x)}, mc.cores = 4)
bird.summary <- mclapply(bird.frame, function(x) {
  as.data.table(x)[, c(mean = lapply(.SD, mean, na.rm=TRUE), 
                       sd = lapply(.SD, sd, na.rm=TRUE)),]
}, mc.cores = 4)
bird.summary.combined <- cbind(data.table(SISID = names(bird.points)),
                                    do.call(rbind, bird.summary))

#Combine with original data 
bird.data.recombined <- bird.summary.combined %>% 
  as.data.frame() %>%
  mutate(SISID = as.integer(SISID)) %>%
  left_join(bird.ranges %>% 
              st_drop_geometry() %>%
              select(SISID, SCINAME, Range.Size) %>%
              as.data.frame())
saveRDS(bird.data.recombined, "data/Intermediate_data/bird.data.recombined.rds")
} else {
bird.data.recombined <- readRDS("data/intermediate_data/bird.data.recombined.rds")
}
```

##R Session Info

```{r}
sessionInfo() %>% pander
```

